### PR TITLE
PWA first tutorial: first 3 parts

### PR DESCRIPTION
**Pull request by estelle**
Set up for intro level PWA tutorial including:

* static HTML and CSS
* JavaScript functionality - fully functional app, but not a PWA
* Manifest file: including appearance, definition, and iconography

Addition sections are currently being worked on.

### CODE REVIEW

File: files/en-us/web/progressive_web_apps/tutorials/intro/javascript_functionality/index.md [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+}
+
+renderPastPeriods();
+```
+
+You can try the fully functioning [period tracker web app](https://mdn.github.io/pwa-examples/intro/javascript_functionality) and view the [period tracker web app source code](https://github.com/mdn/pwa-examples/tree/master/intro/javascript_functionality) on GitHub. Yes, it works, but it's not a yet PWA.
+
+## Up next
+
+At it's core, a PWA is a web application that is progressively enhanced to work offline. Now that we have a fully functional web application, we add the features required to convert into a PWA, including the [manifest file](/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro/Manifest_file), [secure connection](/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro/Secure), and [service worker](/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro/Service_worker), we have to make this application work.
```
Comment (captainbrosset) [inline files/en-us/web/progressive_web_apps/tutorials/intro/javascript_functionality/index.md:347]:
```suggestion
At its core, a PWA is a web application that can be installed is progressively enhanced to work offline. Now that we have a fully functional web application, we add the features required to convert it to a PWA, including the [manifest file](/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro/Manifest_file), [secure connection](/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro/Secure), and [service worker](/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro/Service_worker).
```

File: files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+slug: Web/Progressive_web_apps/Tutorials/Intro/Manifest_file
+---
+
+{{PreviousMenuNext("Web/Progressive_web_apps/Tutorials/Intro/JavaScript_functionality", "Web/Progressive_web_apps/Tutorials/Intro/Secure", "Web/Progressive_web_apps/Tutorials/Intro")}}
+
+{{PWASidebar}}
+
+A PWA manifest file is a JSON file that provides information about the features of that app to make it look and behave like a native app when installed on the user's device. The manifest contains metadata for your app, including its name, icons, and presentational directives.
+
+The manifest file contains information about the PWA's identity, presentation, and iconography. While the only requirement is to include a name using the `name` or `short_name` key, you should also include icon information, the URL to be opened when the PWA is launched, and type of application viewport in which the PWA should be viewed.
```
Comment (captainbrosset) [inline files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:12]:
```suggestion
While the only requirement is to include a name using the `name` or `short_name` key, you should also include icon information, the URL to be opened when the PWA is launched, and type of application viewport in which the PWA should be viewed.
```

This first sentence seems like a repeat of the previous one.

Comment (captainbrosset) [inline files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:12]: It depends what you mean by requirement. Will wrote a guide about PWA installation and it lists the manifest members that are required to install a PWA on a device: https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps/Guides/Making_PWAs_installable#required_manifest_members

File: files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+  "icons": [
+    {
+      "src": "icon-512.png",
+      "sizes": "512x512"
+    }
+  ]
+}
+```
+
+Before saving the manifest file and linking to it from our HTML file, let's discuss we can develop a still brief but more informative JSON object to define the identity, presentation, and iconography of the PWA. Yes, the above would work, but let's discuss a few more keys that enable manifest files to better define the appearance of our PWA.
```
Comment (captainbrosset) [inline files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:28]:
```suggestion
Before saving the manifest file and linking to it from our HTML file, we can develop a still brief but more informative JSON object to define the identity, presentation, and iconography of the PWA. Yes, the above would work, but let's discuss a few more members that enable manifest files to better define the appearance of our PWA.
```

File: files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+## App identity
+
+To identity your PWA the JSON must include a `name` or `short_name` key, or both, to define the PWA name. It can also include a `description`.
```
Comment (captainbrosset) [inline files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:32]:
```suggestion
To identity your PWA the JSON must include a `name` or `short_name` member, or both, to define the PWA name. It can also include a `description`.
```

Top-level properties of a web app manifest are called members. We've been using that name in other PWA docs on MDN. It'd be good to stick to it consistently if we can.

**Review by captainbrosset — COMMENTED** [review]
I'm not fully done with the review, but since I'm done with my day, I thought I would post my comments now, giving you some time to take a look at them, and I'll then go back to the review tomorrow, or a bit later this week.
Thanks for the work on this so far, this looks really good.

File: files/en-us/web/progressive_web_apps/tutorials/intro/html_and_css/index.md [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+  <h2>Past periods</h2>
+  <ul>
+    <li>From 01/01/2024 to 01/06/2024</li>
+    <li>From 01/29/2024 to 02/04/2024</li>
+  </ul>
+  -->
+</section>
+```
+
+You can try the [static period tracker shell](https://mdn.github.io/pwa-examples/intro/html_and_css) and view the [period tracker HTML and CSS source code](https://github.com/mdn/pwa-examples/tree/master/intro/html_and_css) on GitHub.
```
[the remark this answers is not in the document]
Comment (estelle) [reply files/en-us/web/progressive_web_apps/tutorials/intro/html_and_css/index.md:234]: good catch!

File: files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+slug: Web/Progressive_web_apps/Tutorials/Intro/Manifest_file
+---
+
+{{PreviousMenuNext("Web/Progressive_web_apps/Tutorials/Intro/JavaScript_functionality", "Web/Progressive_web_apps/Tutorials/Intro/Secure", "Web/Progressive_web_apps/Tutorials/Intro")}}
+
+{{PWASidebar}}
+
+A PWA manifest file is a JSON file that provides information about the features of that app to make it look and behave like a native app when installed on the user's device. The manifest contains metadata for your app, including its name, icons, and presentational directives.
+
+While according to the spec, all of the manifest members, or keys, are optional, some are browsers, operating systems, and app distributors have requirements, with a manifest needing to include specific keys for a web app to be a PWA. By including a name using the `name` or `short_name` key, icons that meet the minimum requirements, the URL to be opened when the PWA is launched, and type of application viewport in which the PWA should be viewed, your app will meet the manifest requirements of a PWA.
```
Comment (captainbrosset) [inline files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:12]:
```suggestion
While according to the spec, all of the manifest members, or keys, are optional, some browsers, operating systems, and app distributors have requirements to include specific keys for a web app to be a PWA. By including a name using the `name` or `short_name` key, icons that meet the minimum requirements, the URL to be opened when the PWA is launched, and type of application viewport in which the PWA should be viewed, your app will meet the manifest requirements of a PWA.
```

Feel free to change this suggestion, but the original sentence sounded a bit odd to me.

File: files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+
+A minimalist manifest file for our menstrual cycle tracking app could look like this:
```
Comment (captainbrosset) [inline files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:14]: I would be good to link to Will's article about the required members, since that's the place that's most likely to be up to date.

File: files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+{
+  "short_name": "CT",
+  "start_url" : "/",
+  "icons": [
+    {
+      "src": "icon-512.png",
+      "sizes": "512x512"
+    }
+  ],
+  "display": "browser"
```
Comment (captainbrosset) [inline files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:26]: I think that, today, on all OS and devices, the `display:browser` does **not** make a web app a PWA. To be tested, but I think browsers won't offer the button to install the app if this display mode is used. So, even if you have a section about this later, it might make sense to use the `standalone` display mode here instead. Just for illustration purposes.

File: files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+The appearance, or presentation, of a PWA's installed and offline experiences are defined in the manifest. Presentation manifest include the `start_url` and `display` keys, and keys which can be used to [customize your app colors](/en-US/docs/Web/Progressive_web_apps/How_to/Customize_your_app_colors), including `theme_color` and `background_color`.
+
+- [`start_url`](/en-US/docs/Web/Manifest/start_url)
+
+  - : The start page when a user launches the PWA.
+
+- [`display`](/en-US/docs/Web/Manifest/display)
+  - : Controls the app's display mode including `fullscreen`, `standalone`, which displays the [PWA as a standalone application](/en-US/docs/Web/Progressive_web_apps/How_to/Create_a_standalone_app), `minimal-ui`, which is similar to a standalone view but with UI elements for controlling navigation, and `browser`, which opens the app in a regular browser view.
+
+There is also an `orientation` key to define the PWA's default orientation as `portrait` or `landscape`. As our app works either way, we'll omit this key.
```
Comment (captainbrosset) [inline files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:71]: I would personally prefer if we used `member` instead of `key` throughout this article, since that's how we refer to this in other articles we've written. Consistency should help readers understand what these things are. Additionally, this is also the spec vocabulary.

File: files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+- [`background_color`](/en-US/docs/Web/Manifest/background_color)
+  - : A placeholder color to be displayed as the [background of the app](/en-US/docs/Web/Progressive_web_apps/How_to/Customize_your_app_colors#customize_the_app_window_background_color) until the CSS is loaded. To create a smooth transition between app launch and load, it is recommended to use the [`<color>`](/en-US/docs/Web/CSS/color_value) declared as the app's [`background-color`](/en-US/docs/Web/CSS/background-color) color.
+
+### Task
+
+Add presentation definitions to the manifest file you began creating in the previous task.
+
+### Example solution
+
+As the example application is a single page, we can use `"/"`, or omit the key altogether. For that same reason, we can display cycleTracker without browser tools, setting the `display` to `standalone`.
```
Comment (captainbrosset) [inline files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:86]:
```suggestion
As the example application is a single page, we can use `"/"` as the `start_url`, or omit the member altogether. For that same reason, we can display the app without the browser UI by setting the `display` to `standalone`.
```

File: files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+
+In [our CSS](/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro/HTML_and_CSS#css_file), the `background-color: #efe;` is set on the `body` element selector. We use `#eeffee` to ensure a smooth transition from placeholder appearance to app load.
+}
```
Comment (captainbrosset) [inline files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:89]:
```suggestion
```

File: files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+  "start_url": "/",
+  "theme_color": "#eeffee",
+  "background_color": "#eeffee",
+  "display": "standalone"
+}
+```
+
+## App iconography
+
+PWA icons help users identify your app, make it more visually appealing, and improve discoverability. The PWA app icon appears on home screens, app launchers, browser address bars, in app store search results: When users search for apps in the app store, the PWA icon will be displayed in the search results. The size of the rendered icon and the file requirements varies depending on where it is displayed and by whom. The manifest is where you define your images.
```
Comment (captainbrosset) [inline files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:105]:
```suggestion
PWA icons help users identify your app, make it more visually appealing, and improve discoverability. The PWA app icon appears on home screens, app launchers, or app store search results. The size of the rendered icon and the file requirements varies depending on where it is displayed and by whom. The manifest is where you define your images.
```

Just a suggestion to remove the address bar because the favicon is used there, not the manifest icon (afaik).

File: files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+
+Within the manifest JSON object, the `icons` key specifies an array of one or more icon objects for use in different contexts, each with a `src` and `sizes` key, and optional `type` and `purpose` keys. Each icon object's' `src` list the source of a single image file. The `sizes` key provides a list of space-separated sizes for which that particular image should be used or the keyword `any`; the value is the same as the {{HTMLElement("link")}} element's [`sizes`](/en-US/docs/Web/HTML/Element/link#sizes) attribute. The `type` key lists the image's MIME type.
+
+```js
+{
+  "name": "MyApp",
+  "icons: [
```
Comment (captainbrosset) [inline files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:112]:
```suggestion
  "icons": [
```

File: files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+    },
+    {
+      "src": "icon/scalable.svg",
+      "sizes": "any"
+    }
+  ]
+}
+```
+
+The `purpose` keys should be set to `maskable` defining [icona as adaptive](https://web.dev/maskable-icon/).
```
Comment (captainbrosset) [inline files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:134]: Hm, I'm having trouble parsing this sentence. 
Also, not all icons need to be maskable. And we might want to link to https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps/How_to/Define_app_icons#support_masking for more information on this topic.

File: files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+
+All icons should have the same look and feel to ensure users recognize your PWA, but the larger the icon, the greater the detail it can contain. All icon files are squares. Include safe zones so that when the image is masked by the operating system, it renders okay as a circle.
```
Comment (captainbrosset) [inline files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:136]: It might make sense to expand a little bit on what the safe zone is. We can borrow some of the text from that "support masking" section I mentioned before if needed.

File: files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+}
+```
+
+## Adding the manifest to the app
+
+You now have a fully usable manifest file. Time to save it and link to it from our HTML file.
+
+The manifest file extension can be the specification suggestion `.webappmanifest`. However, being a JSON file, it is most commonly saved with the browser-supported `.json` extension.
+
+PWAs require a manifest file. We have a fully functional app, but it's not yet a PWA. We have to add the manifest the app. The web manifest is an external JSON file. To include the external JSON resource, the `rel="manifest"` is used. The `href` attribute of the `<link>` points to the location of the resource.
```
Comment (captainbrosset) [inline files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:203]:
```suggestion
PWAs require a manifest file to be linked from the app's HTML document. We have a fully functional app, but it's not yet a PWA because it doesn't link to our external manifest JSON file yet. To include the external JSON resource, we use the `<link>` element, with the `rel="manifest"` attribute, and setting the `href` attribute to the location of the resource.
```

Just a suggestion to improve the flow of this paragraph, which seemed a bit hard to follow at first read.

File: files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+
+Save the manifest file that you have created in the steps above, then link to it from the `index.html` file.
+
+#### Example solution
+
+View the [`cycletracker.manifest` file](https://mdn.github.io/pwa-examples/cycletracker/Manifest_file/cycletracker.manifest) and view the [project source code](https://github.com/mdn/pwa-examples/tree/master/cycletracker/Manifest_file) on GitHub.
+
+## Up next
+
+Now that we have a PWA manifest file, let's [create a secure connection](/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro/Secure). While no frameworks are needed to create a PWA, we will use X to create a secure connection; a PWA requirement.
```
Comment (captainbrosset) [inline files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:227]: Curious what you have in mind with framework X in this sentence?
Note that when you're working on localhost (as most devs will do), there is no need for a secure connection. The browser will allow installing the PWA on http if the domain is localhost.

**Review by captainbrosset — COMMENTED** [review]
Some more comments. I hope they're useful. Thanks for the work on this!

File: files/en-us/web/progressive_web_apps/tutorials/intro/html_and_css/html.jpg
[the remark this answers is not in the document]
Comment (estelle) [reply files/en-us/web/progressive_web_apps/tutorials/intro/html_and_css/html.jpg:1]: i added a new image: screen shot from MS Edge ;)

File: files/en-us/web/progressive_web_apps/tutorials/intro/html_and_css/index.md [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+    <script src="app.js" defer></script>
+  </body>
+</html>
+```
+
+Copy this HTML and save it in a file called `index.html`.
+
+## HTML content
+
+Even if the HTML in `index.html` is familiar to you, we recommend reading thru this section before adding the [placeholder data](en-US/Docs/Web/Progressive_web_apps/Tutorials/Intro/HTML_and_CSS#placeholder_text) and [`styles.css`](en-US/Docs/Web/Progressive_web_apps/Tutorials/Intro/HTML_and_CSS#css-file), and creating `app.js`, the [application's JavaScript](/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro/JavaScript_functionality) that makes this web page function.
```
[the remark this answers is not in the document]
Comment (estelle) [reply files/en-us/web/progressive_web_apps/tutorials/intro/html_and_css/index.md:61]: this is kind of the table of contents for this page, but reworded it to "adding some temporary hard-coded data"

File: files/en-us/web/progressive_web_apps/tutorials/intro/javascript_functionality/index.md [hunk 19]
```diff
@@ -0,0 +1,351 @@
...
+Note the last line in the body links to the `app.js` script. This is the script we are creating in this section. In this lesson, we will be writing client-side JavaScript to capture form submissions, locally store submitted data, and populate the past periods section.
+
+At the end of this lesson, you will have a fully functional app. In future lessons, we will progressively enhance the app to create an fully installable PWA that works even when the user is offline.
+
+## Javascript task
+
+When a user visits the page, we check if they have existing data stored in local storage. The first time a user visits the page, there won't be any data. When a new user selects two dates and submits the form, we need to:
+
+1. Create a "`<h2>Past periods</h2>`" header
+2. Create an {{HTMLelement("ul")}}
```
[the remark this answers is not in the document]
Comment (estelle) [reply files/en-us/web/progressive_web_apps/tutorials/intro/javascript_functionality/index.md:23]: i debated doing this. It was easier to add a line of JS then to explain the UX design requirements of empty data along with the CSS for the two different experiences. (the "empty" message would not be a list item semantically)

File: files/en-us/web/progressive_web_apps/tutorials/intro/javascript_functionality/index.md [hunk 20]
```diff
@@ -0,0 +1,351 @@
...
+
+We check if the dates are invalid. We do minimal error checking. We make sure neither date is null, which the `required` attribute should prevent from happening. We also check that the start date is not greater than the end date. If there is an error, we clear the form.
+
+```javascript
+function checkDatesInvalid(startDate, endDate) {
+  // Check that end date is after start date and neither is null.
+  if (!startDate || !endDate || startDate > endDate) {
+    // For now, we clear the dates if either
+    // or both are invalid
+    newPeriodFormEl.reset();
```
[the remark this answers is not in the document]
Comment (estelle) [reply files/en-us/web/progressive_web_apps/tutorials/intro/javascript_functionality/index.md]: I understand and agree with what you are saying. If we were creating a robust app, this would not reset the form, but rather do the steps explained in the comments. That is why this was created as a separate function -- for the comments, not the current code. I reordered the comments in the function to make this clearer.

**Comment by estelle** [comment]
> Some more comments. I hope they're useful. Thanks for the work on this!

Thanks for the excellent feedback. I addressed all the issues (I think), resolving those where I followed your suggestions, which were all but 2: 1) the checkDateInvalid is indeed not a good function. I kept it as a place to hold comments to explain what would be required for a better function with suggestions for better validation, and 2) i kept the h2 in the JS, keeping the original div empty when there is no content, to avoid having to use an `li` when a `p` is appropriate, to not to have to explain UI of empty data, and to avoid have to write and explain CSS associated with that UI.

**Comment by captainbrosset** [comment]
@estelle the content looks great, but I'm confused about one thing that I'd like to understand before approving this: when I go to the preview site for this PR, I can see a bunch of content that's not in the PR, like this intro page: https://pr26891.content.dev.mdn.mozit.cloud/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro
Or the changes you made to the tutorial main page: https://pr26891.content.dev.mdn.mozit.cloud/en-US/docs/Web/Progressive_web_apps/Tutorials

Do you know where this content comes from?

I also see a lot of red links in the preview and I don't know if you intend to add the remaining pages in this PR or later.
Personally, I'm happy for this to be merge, knowing that it's already been a lot of work, and then merging follow-up PRs to finish everything up.

Let me know.

**Comment by estelle** [comment]
There are two PRs that are not ready for merge: 

the [/en-US/docs/Web/Progressive_web_apps/Tutorials](https://pr26891.content.dev.mdn.mozit.cloud/en-US/docs/Web/Progressive_web_apps/Tutorials) PR can be reviewed and approved, but should not be published until the tutorial is complete. That's the landing page for all the PWA tutorials. 

The [/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro](https://pr26891.content.dev.mdn.mozit.cloud/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro) is the landing page for this tutorial. It won't be finished until the tutorial is finished as the overview of something keeps getting edited until that something is complete. That is in draft mode and is there for reviewers to understand the scope of the project and what missing sections are actually TBW.

I broke the project into a few PRs because, as you'll note, this PR already has 100 or so commits and over 50 comments. A whole project PR would be overwhelming.


> 
> Do you know where this content comes from?
> 
> I also see a lot of red links in the preview and I don't know if you intend to add the remaining pages in this PR or later. Personally, I'm happy for this to be merge, knowing that it's already been a lot of work, and then merging follow-up PRs to finish everything up.
> 
> Let me know.
